lib/spear/connection/request.ex

```diff
@@ -0,0 +1,194 @@
+defmodule Spear.Connection.Request do
+  @moduledoc false
+
+  # a struct representing a stream-able request
+
+  @type t :: %{
+          continuation: Enumerable.continuation(),
+          request_ref: Mint.Types.request_ref(),
+          buffer: binary(),
+          from: GenServer.from(),
+          response: Spear.Connection.Response.t(),
+          status: :streaming | :done
+        }
+
+  defstruct [:continuation, :buffer, :request_ref, :from, :response, :status]
+
+  def new(event_stream, request_ref, from) do
+    reducer = &reduce_with_suspend/2
+    stream = Stream.map(event_stream, &Spear.Request.to_wire_data/1)
+    continuation = &Enumerable.reduce(stream, &1, reducer)
+
+    %__MODULE__{
+      continuation: continuation,
+      buffer: <<>>,
+      request_ref: request_ref,
+      from: from,
+      response: %Spear.Connection.Response{},
+      status: :streaming
+    }
+  end
+
+  defp reduce_with_suspend(
+         {message, message_size},
+         {message_buffer, message_buffer_size, max_size}
+       )
+       when message_size + message_buffer_size > max_size do
+    {:suspend,
+     {[{message, message_size} | message_buffer], message_size + message_buffer_size, max_size}}
+  end
+
+  defp reduce_with_suspend(
+         {message, message_size},
+         {message_buffer, message_buffer_size, max_size}
+       ) do
+    {:cont,
+     {[{message, message_size} | message_buffer], message_size + message_buffer_size, max_size}}
+  end
+
+  @spec emit_messages(%Spear.Connection{}, %__MODULE__{}) ::
+          {:ok, %Spear.Connection{}} | {:error, %Spear.Connection{}, reason :: any()}
+  def emit_messages(state, %__MODULE__{buffer: <<>>, continuation: continuation} = request) do
+    smallest_window = get_smallest_window(state.conn, request.request_ref)
+
+    {:cont, {[], 0, smallest_window}}
+    |> continuation.()
+    |> handle_contination(state, request)
+  end
+
+  def emit_messages(
+        state,
+        %__MODULE__{buffer: buffer, continuation: continuation} = request
+      ) do
+    smallest_window = get_smallest_window(state.conn, request.request_ref)
+
+    case buffer do
+      <<bytes_to_send::binary-size(smallest_window), rest::binary>> ->
+        state
+        |> put_request(%__MODULE__{request | buffer: rest})
+        |> stream_messages(
+          request.request_ref,
+          [{bytes_to_send, smallest_window}]
+        )
+
+      ^buffer ->
+        # buffer is small enough to be sent in one go
+        # so we resume the happy path of cramming as many messages as possible
+        # into frames
+        buffer_size = byte_size(buffer)
+
+        {:cont, {[{buffer, buffer_size}], buffer_size, smallest_window}}
+        |> continuation.()
+        |> handle_contination(state, request)
+    end
+  end
+
+  defp handle_contination(
+         {finished, {message_buffer, _buffer_size, _max_size}},
+         state,
+         request
+       )
+       when finished in [:done, :halted] do
```

Comment:
I wasn't clear on exactly what the difference is here. Best I can tell

- `:done`: finite enumerables like lists
- `:halted`: streams